### PR TITLE
Places fix score and quality with autocomplete integration test

### PR DESCRIPTION
**Pull request by kadhikari**

### CODE REVIEW

File: source/autocomplete/autocomplete.cpp [hunk 1]
```diff
@@ -66,9 +66,11 @@ void compute_score_stop_point(type::PT_Data &pt_data, georef::GeoRef &georef) {
     }
 }
 
-void compute_score_stop_area(type::PT_Data & pt_data){
+void compute_score_stop_area(type::PT_Data & pt_data, georef::GeoRef &georef){
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: GeoRef (and PT_Data) should be taken const

File: source/autocomplete/autocomplete.cpp [hunk 2]
```diff
@@ -76,13 +78,29 @@ void compute_score_stop_area(type::PT_Data & pt_data){
     //Ajust the score of each stop_area from 0 to 100 using maximum score (max_score)
     if (max_score > 0){
         for (auto & it : pt_data.stop_area_autocomplete.word_quality_list){
-            it.second.score = (pt_data.stop_areas[it.first]->stop_point_list.size() * 100)/max_score;
+            for (navitia::georef::Admin* admin : pt_data.stop_areas[it.first]->admin_list){
+                if(admin->level == 8){
+                    admin_score = georef.fl_admin.word_quality_list.at(admin->idx).score;
+                }
+            }
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: I'd prefere a small static function like that:

``` c++
static size_t admin_score(const std::vector<navitia::georef::Admin*>& admins, const georef::GeoRef &georef) {
    for (const auto* admin: admins) {
        if (admin->level == 8) return georef.fl_admin.word_quality_list.at(admin->idx).score;
    }
    return 0;
}
```

and then, here

``` c++
const size_t admin_score = admin_score(pt_data.stop_areas[it.first]->admin_list, georef);
```

(and removing current admin_score declaration)

(not tested, may not compile)

File: source/autocomplete/autocomplete.cpp [hunk 3]
```diff
@@ -91,14 +109,10 @@ void compute_score_admin(type::PT_Data &pt_data, georef::GeoRef &georef) {
...
-    //Calculate max_score of all admins
-    for (auto it = georef.fl_admin.word_quality_list.begin(); it != georef.fl_admin.word_quality_list.end(); ++it){
-        max_score = ((it->second).score > max_score)?(it->second).score:max_score;
-    }
 
-    //Ajust the score of each admin using max_score.
+    //Ajust the score of each admin using natural logarithm as : log(n+2)*10
     for (auto it = georef.fl_admin.word_quality_list.begin(); it != georef.fl_admin.word_quality_list.end(); ++it){
-        (it->second).score = max_score == 0 ? 0 : ((it->second).score * 100)/max_score;
+        (it->second).score = log((it->second).score + 2) * 10;
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: You can write `it->second.score` directly if you prefere.

File: source/autocomplete/autocomplete.h [hunk 4]
```diff
@@ -297,13 +300,27 @@ struct Autocomplete
...
 
-    std::vector<fl_quality> sort_and_truncate(std::vector<fl_quality> input, size_t nbmax) const {
+    std::vector<fl_quality> sort_and_truncate_by_score(std::vector<fl_quality> input, size_t nbmax) const {
+        typename std::vector<fl_quality>::iterator middle_iterator;
+        if(nbmax < input.size())
+            middle_iterator = input.begin() + nbmax;
+        else
+            middle_iterator = input.end();
+        std::partial_sort(input.begin(), middle_iterator, input.end(),
+                          [&](fl_quality a, fl_quality b){return (a.score > b.score);});
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.h]: In a lambda, the & between the brackets say that you can take any value in the environment by reference. As you do not need any environment here, you can omit the & like that:

``` c++
[](fl_quality a, fl_quality b) {return a.score > b.score;}
```

if you prefere.

File: source/autocomplete/autocomplete_api.cpp [hunk 5]
```diff
@@ -52,6 +52,7 @@ void fill_pb_pt_object(T* nav_object, pbnavitia::PtObject* place,
     place->set_uri(nav_object->uri);
     place->set_quality(100);
     place->set_embedded_type(pb_type);
+
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete_api.cpp]: extra added line

File: source/autocomplete/autocomplete_api.cpp [hunk 6]
```diff
@@ -341,12 +321,37 @@ pbnavitia::Response autocomplete(const std::string &q,
...
+            return a.quality() > b.quality();
+        }
+    };
+
+    if (search_type != 0) {
+        nbmax = nbmax_temp;
+        auto mutable_places = pb_response.mutable_places();
+        int result_size = std::min(nbmax, mutable_places->size());
+        std::partial_sort(mutable_places->begin(),mutable_places->begin() + result_size,
+                          mutable_places->end(),compare_by_quality);
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete_api.cpp]: Maybe we can write a sort_and_truncate method in utils, and use it everywere (or maybe not)

**Comment by TeXitoi** [comment]
globally, it seems great to me, with lots of tests! :)

File: source/autocomplete/autocomplete.h [hunk 7]
```diff
@@ -297,13 +300,27 @@ struct Autocomplete
...
+
+    std::vector<fl_quality> sort_and_truncate_by_quality(std::vector<fl_quality> input, size_t nbmax) const {
         typename std::vector<fl_quality>::iterator middle_iterator;
         if(nbmax < input.size())
             middle_iterator = input.begin() + nbmax;
         else
             middle_iterator = input.end();
-        std::partial_sort(input.begin(), middle_iterator, input.end());
+        std::partial_sort(input.begin(), middle_iterator, input.end(),
+                          [&](fl_quality a, fl_quality b){return (a.quality > b.quality);});
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.h]: Maybe we can factorize these two methods together:

``` c++
template<typename Cmp>
static std::vector<fl_quality> sort_and_truncate(std::vector<fl_quality> input, size_t nbmax, Cmp cmp) {
    typename std::vector<fl_quality>::iterator middle_iterator;
    if(nbmax < input.size())
        middle_iterator = input.begin() + nbmax;
    else
        middle_iterator = input.end();
    std::partial_sort(input.begin(), middle_iterator, input.end(), cmp);
    if (input.size() > nbmax){input.resize(nbmax);}
    return input;
}
```

And then, our 2 methods:

``` c++
std::vector<fl_quality> sort_and_truncate_by_quality(std::vector<fl_quality> input, size_t nbmax) const {
    return sort_and_truncate(input, nbmax, [](const fl_quality& a, const fl_quality& b){return a.quality > b.quality;)
}
...
```

File: source/autocomplete/autocomplete.cpp [hunk 8]
```diff
@@ -60,15 +60,24 @@ void compute_score_stop_point(type::PT_Data &pt_data, georef::GeoRef &georef) {
...
+        if (admin->level == 8) return georef.fl_admin.word_quality_list.at(admin->idx).score;
+    }
+    return 0;
+}
+
+void compute_score_stop_area(type::PT_Data & pt_data, const georef::GeoRef &georef){
+    //The scocre of each admin(level 8) is attributed to all its stop_areas also
     //Find the stop-point count in all stop_areas and keep the highest;
     size_t max_score = 0;
+    //size_t admin_score = 0;
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.cpp]: to remove

File: source/autocomplete/autocomplete.h (hunk 7 above)
Comment (TeXitoi) [inline source/autocomplete/autocomplete.h]: This remark was forgotten, but we can do without.

File: source/autocomplete/autocomplete.h [hunk 9]
```diff
@@ -297,13 +300,27 @@ struct Autocomplete
...
+        if(nbmax < input.size())
+            middle_iterator = input.begin() + nbmax;
+        else
+            middle_iterator = input.end();
+        std::partial_sort(input.begin(), middle_iterator, input.end(),
+                          [](fl_quality a, fl_quality b){return a.score > b.score;});
+
+        if (input.size() > nbmax){input.resize(nbmax);}
+        return input;
+    }
```
Comment (TeXitoi) [inline source/autocomplete/autocomplete.h:305]: I paste here a comment that have been removed accidentally by github:

Maybe we can factorize these two methods together:

``` c++
template<typename Cmp>
static std::vector<fl_quality> sort_and_truncate(std::vector<fl_quality> input, size_t nbmax, Cmp cmp) {
    typename std::vector<fl_quality>::iterator middle_iterator;
    if(nbmax < input.size())
        middle_iterator = input.begin() + nbmax;
    else
        middle_iterator = input.end();
    std::partial_sort(input.begin(), middle_iterator, input.end(), cmp);
    if (input.size() > nbmax){input.resize(nbmax);}
    return input;
}
```

And then, our 2 methods:

``` c++
std::vector<fl_quality> sort_and_truncate_by_quality(std::vector<fl_quality> input, size_t nbmax) const {
    return sort_and_truncate(input, nbmax, [](const fl_quality& a, const fl_quality& b){return a.quality > b.quality;)
}
...
```

File: source/autocomplete/autocomplete_api.cpp (hunk 6 above)
Comment (antoine-de) [inline source/autocomplete/autocomplete_api.cpp]: I agree with that

**Comment by TeXitoi** [comment]
OK when jenkins is OK and sort_and_truncate is in utils and used.

sort_and_truncate in utils/function.h:

``` c++
template <typename Vector, typename Cmp>
Vector sort_and_truncate(Vector input, size_t nbmax, Cmp cmp) {
    typename Vector::iterator middle_iterator;
    if (nbmax < input.size())
        middle_iterator = input.begin() + nbmax;
    else
        middle_iterator = input.end();
    std::partial_sort(input.begin(), middle_iterator, input.end(), cmp);
    if (input.size() > nbmax) input.resize(nbmax);
    return std::move(input);
}
```

**Comment by antoine-de** [comment]
Jenkins, test this please

**Pull request merged**